### PR TITLE
fix: prevent lint all files

### LINKED ISSUE #75: All modules are linted in watch mode after changing only a single file
- Operating System: macOS 11.1
- Node Version: 15.7.0
- NPM Version: yarn berry PnP
- webpack Version: 4.46.0
- eslint-webpack-plugin Version: 2.5.0

### Expected Behavior

Only lint modules that were rebuilt after changing a file.

### Actual Behavior

Every module is linted after changing a file.

### Code

```js
new ESLintPlugin({
  extensions: ['js', 'vue'],
})
```

### How Do We Reproduce?

It seems there was a recent change to which hooks were used to lint the files. Before `succeedModule` was used which was only called with the rebuilt modules after a file change, but now `finishModules` is used which passes every module in the compilation, so all files get linted.

### PR DESCRIPTION
**Pull request by ricardogobbosouza**
This PR contains a:

- [x] **bugfix**
- [ ] new **feature**
- [ ] **code refactor**
- [ ] **test update** 
- [ ] **typo fix**
- [ ] **metadata update**

### Motivation / Use-Case

The goal is to group the changed files and execute lint only at the end

`succeedModule` its called for every file and watch mode just file changed
`finishModules` is called with all modules and is responsible for linting added files

Resolve #75

### CODE REVIEW

**Review by alexander-akait — COMMENTED** [review]
I think we need to look at better hook and logic here, but right now is good solution, let's return to this in future

Will be great to add test using `mock`

File: src/index.js [hunk 1]
```diff
@@ -95,27 +95,28 @@ class ESLintWebpackPlugin {
...
+          if (
+            file &&
+            !files.includes(file) &&
+            isMatch(file, wanted) &&
+            !isMatch(file, exclude)
+          ) {
+            files.push(file);
           }
         }
+      });
```
Comment (WilkinWendy) [inline src/index.js:116]: sorry
I can't understand why this fix can resolve current issue.  
In my opinion,this just collect the files in another time.

Comment (watjurk) [reply src/index.js:116]: agreed

Comment (ricardogobbosouza) [reply src/index.js:116]: The goal is to group the changed files and execute lint only at the end

`succeedModule` its called for every file and watch mode just file changed
`finishModules` is called with all modules and is responsible for linting added files

Comment (WilkinWendy) [reply src/index.js:116]: When hot-update happens,i works,But when the project is started ,all files will be linted
https://github.com/webpack-contrib/eslint-webpack-plugin/issues/75#issuecomment-774022336

Comment (ricardogobbosouza) [reply src/index.js:116]: For this you can use `lintDirtyModulesOnly`

**Comment by watjurk** [comment]
unfortunately this fix won't work after merging `fix-cache-result-storage` because files that aren't changed won't be linted:
-index.js (imports lintError.js)
-lintError.js

webpack reports errors from lintError.js
now let's change index js (still imports lintError.js)
webpack doesn't report errors from lintError.js because `succeedModule` in watch mode (after first run) is called only for changed files

I would suggest not merging this PR and creating global cache in `linter.js` and invalidating it with `compiler.modifiedFiles` and `compiler.removedFiles`.
Example implementation: https://github.com/watjurk/eslint-webpack-plugin/tree/fix-cache-result-storage

**Comment by ricardogobbosouza** [comment]
This will only be possible in v3 because `compiler.modifiedFiles` does not exist in webpack 4

**Comment by ricardogobbosouza** [comment]
In v3 version support for webpack 4 will be removed

**Comment by watjurk** [comment]
when v3 will be available, as beta or "@next" ?

**Comment by watjurk** [comment]
I've learned about `compiler.hooks.invalid`,  I think this is the solution here. I updated my implementation accordingly.

**Review by WilkinWendy — APPROVED** [review]

**Comment by ricardogobbosouza** [comment]
It looks good to me :)

**Comment by watjurk** [comment]
Should I create PR to this repo ?

**Comment by ricardogobbosouza** [comment]
Do you want to create a pr or do you prefer me to do it?

**Comment by watjurk** [comment]
I think I prefer you to do this, maybe some day...

**Comment by WilkinWendy** [comment]
> compiler.hooks.invalid

why `compiler.hooks.invalid` is ok? i find it's `fs.watcher.once('change')`

**Comment by watjurk** [comment]
Sorry but I don't understand your point, could you explain ?

**Comment by WilkinWendy** [comment]
> I've learned about `compiler.hooks.invalid`, I think this is the solution here. I updated my implementation accordingly.

I just don't known the reason, why `compiler.hooks.invalid` is the solution.
Will `invalid` be triggered when any file be deleted?

**Comment by ricardogobbosouza** [comment]
According to the documentation, the invalid hook is called to the changed or removed file
https://webpack.js.org/api/compiler-hooks/#invalid

**Comment by WilkinWendy** [comment]
> According to the documentation, the invalid hook is called to the changed or removed file
> https://webpack.js.org/api/compiler-hooks/#invalid

image
image
but the source code of webpack v4 confused me

**Comment by watjurk** [comment]
> > According to the documentation, the invalid hook is called to the changed or removed file
> > https://webpack.js.org/api/compiler-hooks/#invalid
> > https://github.com/webpack/webpack/blob/fd056e27f477e364c16d1b09b39905fce28c13ca/lib/Watching.js#L252
> > image
> > image
> > but the source code of webpack v4 confused me

Search for _done method in the Watching.js, and you will find:
```js
process.nextTick(() => {
  if (!this.closed) {
    this.watch(
      compilation.fileDependencies,
      compilation.contextDependencies,
      compilation.missingDependencies
    );
  }
});
```
as you see this.watch is called every time webpack is done compiling, and this is why this.watcher.once i really "not once".

**Comment by WilkinWendy** [comment]
> Search for _done method in the Watching.js, and you will find:
> 
> ```js
> process.nextTick(() => {
>   if (!this.closed) {
>     this.watch(
>       compilation.fileDependencies,
>       compilation.contextDependencies,
>       compilation.missingDependencies
>     );
>   }
> });
> ```
> 
> as you see this.watch is called every time webpack is done compiling, and this is why this.watcher.once i really "not once".

Thank you ,so the name of this hook called 'invalid' is not strict ,is it ?

**Comment by watjurk** [comment]
What do you mean by saying it's not strict ? It's just called 'invalid' because it's called every time compilation is invalidated (something change), this is how I think about it.

**Comment by WilkinWendy** [comment]
> What do you mean by saying it's not strict ? It's just called 'invalid' because it's called every time compilation is invalidated (something change), this is how I think about it.

yes , you are right.that's it, I have to learn webpack in detail.

**Comment by ricardogobbosouza** [comment]
This PR temporarily solves the performance problem ...
Waiting #79

**Pull request merged**